File: spark/src/main/scala/com/alibaba/graphar/EdgeInfo.scala
package com.alibaba.graphar

import java.io.{File, FileInputStream}
import org.yaml.snakeyaml.Yaml
import org.yaml.snakeyaml.constructor.Constructor
import scala.beans.BeanProperty

class EdgeInfo() {
  @BeanProperty var src_label: String = ""
  @BeanProperty var edge_label: String = ""
  @BeanProperty var dst_label: String = ""
  @BeanProperty var chunk_size: Long = 0
  @BeanProperty var src_chunk_size: Long = 0
  @BeanProperty var dst_chunk_size: Long = 0
  @BeanProperty var directed: Boolean = false
  @BeanProperty var prefix: String = ""
  @BeanProperty var adj_lists = new java.util.ArrayList[AdjList]()
  @BeanProperty var version: String = ""

  def containAdjList(adj_list_type: AdjListType.Value): Boolean = {
    val tot: Int = adj_lists.size
    for ( k <- 0 to tot - 1 ) {
      val adj_list = adj_lists.get(k)
      if (adj_list.getAdjList_type_in_gar == adj_list_type)
        return true
    }
    return false
  }

  def getAdjListPrefix(adj_list_type: AdjListType.Value): String = {
    val tot: Int = adj_lists.size
    for ( k <- 0 to tot - 1 ) {
      val adj_list = adj_lists.get(k)
      if (adj_list.getAdjList_type_in_gar == adj_list_type) {
        var str: String = adj_list.getPrefix
        if (str == "") {
          str = AdjListType.AdjListTypeToString(adj_list_type) + "/"
        }
        return str
      }
    }
    throw new IllegalArgumentException
  }

  def getAdjListFileType(adj_list_type: AdjListType.Value): FileType.Value = {
    val tot: Int = adj_lists.size
    for ( k <- 0 to tot - 1 ) {
      val adj_list = adj_lists.get(k)
      if (adj_list.getAdjList_type_in_gar == adj_list_type) {
        return adj_list.getFile_type_in_gar
      }
    }
    throw new IllegalArgumentException
  }

  def getPropertyGroups(adj_list_type: AdjListType.Value): java.util.ArrayList[PropertyGroup] = {
    val tot: Int = adj_lists.size
    for ( k <- 0 to tot - 1 ) {
      val adj_list = adj_lists.get(k)
      if (adj_list.getAdjList_type_in_gar == adj_list_type) {
        return adj_list.getProperty_groups
      }
    }
    throw new IllegalArgumentException
  }

  def containPropertyGroup(property_group: PropertyGroup, adj_list_type: AdjListType.Value): Boolean = {
    val tot: Int = adj_lists.size
    for ( k <- 0 to tot - 1 ) {
      val adj_list = adj_lists.get(k)
      if (adj_list.getAdjList_type_in_gar == adj_list_type) {
        val property_groups = adj_list.getProperty_groups
        val len: Int = property_groups.size
        for ( i <- 0 to len - 1 ) {
          val pg: PropertyGroup = property_groups.get(i)
          if (pg == property_group) {
            return true
          }
        }
      }
    }
    return false
  }

  def containProperty(property_name: String): Boolean = {
    val tot: Int = adj_lists.size
    for ( k <- 0 to tot - 1 ) {
      val adj_list = adj_lists.get(k)
      val property_groups = adj_list.getProperty_groups
      val len: Int = property_groups.size
      for ( i <- 0 to len - 1 ) {
        val pg: PropertyGroup = property_groups.get(i)
        val properties = pg.getProperties
        val num = properties.size
        for ( j <- 0 to num - 1 ) {
          if (properties.get(j).getName == property_name) {
            return true
          }
        }
      }
    }
    return false
  }

  def getPropertyGroup(property_name: String, adj_list_type: AdjListType.Value): PropertyGroup = {
    val tot: Int = adj_lists.size
    for ( k <- 0 to tot - 1 ) {
      val adj_list = adj_lists.get(k)
      if (adj_list.getAdjList_type_in_gar == adj_list_type) {
        val property_groups = adj_list.getProperty_groups
        val len: Int = property_groups.size
        for ( i <- 0 to len - 1 ) {
          val pg: PropertyGroup = property_groups.get(i)
          val properties = pg.getProperties
          val num = properties.size
          for ( j <- 0 to num - 1 ) {
            if (properties.get(j).getName == property_name) {
              return pg
            }
          }
        }
      }
    }
    throw new IllegalArgumentException
  }

  def getPropertyType(property_name: String): GarType.Value = {
    val tot: Int = adj_lists.size
    for ( k <- 0 to tot - 1 ) {
      val adj_list = adj_lists.get(k)
      val property_groups = adj_list.getProperty_groups
      val len: Int = property_groups.size
      for ( i <- 0 to len - 1 ) {
        val pg: PropertyGroup = property_groups.get(i)
        val properties = pg.getProperties
        val num = properties.size
        for ( j <- 0 to num - 1 ) {
          if (properties.get(j).getName == property_name) {
            return properties.get(j).getData_type_in_gar
          }
        }
      }
    }
    throw new IllegalArgumentException
  }

  def isPrimaryKey(property_name: String): Boolean = {
    val tot: Int = adj_lists.size
    for ( k <- 0 to tot - 1 ) {
      val adj_list = adj_lists.get(k)
      val property_groups = adj_list.getProperty_groups
      val len: Int = property_groups.size
      for ( i <- 0 to len - 1 ) {
        val pg: PropertyGroup = property_groups.get(i)
        val properties = pg.getProperties
        val num = properties.size
        for ( j <- 0 to num - 1 ) {
          if (properties.get(j).getName == property_name) {
            return properties.get(j).getIs_primary
          }
        }
      }
    }
    throw new IllegalArgumentException
  }

  def getPrimaryKey(): String = {
    val tot: Int = adj_lists.size
    for ( k <- 0 to tot - 1 ) {
      val adj_list = adj_lists.get(k)
      val property_groups = adj_list.getProperty_groups
      val len: Int = property_groups.size
      for ( i <- 0 to len - 1 ) {
        val pg: PropertyGroup = property_groups.get(i)
        val properties = pg.getProperties
        val num = properties.size
        for ( j <- 0 to num - 1 ) {
          if (properties.get(j).getIs_primary) {
            return properties.get(j).getName
          }
        }
      }
    }
    return ""
  }

  def isValidated(): Boolean = {
    if (src_label == "" || edge_label == "" || dst_label == "")
      return false
    if (chunk_size <= 0 || src_chunk_size <= 0 || dst_chunk_size <= 0)
      return false
    val tot: Int = adj_lists.size
    for ( k <- 0 to tot - 1 ) {
      val adj_list = adj_lists.get(k)
      val file_type = adj_list.getFile_type_in_gar
      val property_groups = adj_list.getProperty_groups
      val len: Int = property_groups.size
      for ( i <- 0 to len - 1 ) {
        val pg: PropertyGroup = property_groups.get(i)
        val properties = pg.getProperties
        val num = properties.size
        if (num == 0)
          return false
        val pg_file_type = pg.getFile_type_in_gar
      }
    }
    return true
  }

  def getAdjListOffsetFilePath(chunk_index: Long, adj_list_type: AdjListType.Value) : String = {
    if (containAdjList(adj_list_type) == false)
      throw new IllegalArgumentException
    val str: String = prefix + getAdjListPrefix(adj_list_type) + "offset/part" +
           chunk_index.toString() + "/chunk0"
    return str
  }

  def getAdjListOffsetDirPath(adj_list_type: AdjListType.Value) : String = {
    if (containAdjList(adj_list_type) == false)
      throw new IllegalArgumentException
    return prefix + getAdjListPrefix(adj_list_type) + "offset/"
  }

  def getAdjListFilePath(vertex_chunk_index: Long, chunk_index: Long, adj_list_type: AdjListType.Value) : String = {
    var str: String = prefix + getAdjListPrefix(adj_list_type) + "adj_list/part" +
        vertex_chunk_index.toString() + "/chunk" + chunk_index.toString()
    return str
  }

  def getAdjListDirPath(adj_list_type: AdjListType.Value) : String = {
    return prefix + getAdjListPrefix(adj_list_type) + "adj_list/"
  }

  def getPropertyFilePath(property_group: PropertyGroup, adj_list_type: AdjListType.Value, vertex_chunk_index: Long, chunk_index: Long) : String = {
    if (containPropertyGroup(property_group, adj_list_type) == false)
      throw new IllegalArgumentException
    var str: String = property_group.getPrefix
    if (str == "") {
      val properties = property_group.getProperties
      val num = properties.size
      for ( j <- 0 to num - 1 ) {
        if (j > 0)
          str += GeneralParams.regularSeperator
        str += properties.get(j).getName;
      }
      str +=  "/"
    }
    str = prefix + getAdjListPrefix(adj_list_type) + str + "part" +
        vertex_chunk_index.toString() + "/chunk" + chunk_index.toString()
    return str
  }

  def getPropertyDirPath(property_group: PropertyGroup, adj_list_type: AdjListType.Value) : String = {
    if (containPropertyGroup(property_group, adj_list_type) == false)
      throw new IllegalArgumentException
    var str: String = property_group.getPrefix
    if (str == "") {
      val properties = property_group.getProperties
      val num = properties.size
      for ( j <- 0 to num - 1 ) {
        if (j > 0)
          str += GeneralParams.regularSeperator
        str += properties.get(j).getName;
      }
      str +=  "/"
    }
    str = prefix + getAdjListPrefix(adj_list_type) + str
    return str
  }
}

File: maven-projects/info/src/main/java/org/apache/graphar/info/EdgeInfo.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package org.apache.graphar.info;

import java.net.URI;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.Set;
import java.util.function.Function;
import java.util.stream.Collectors;
import java.util.stream.Stream;
import org.apache.graphar.info.type.AdjListType;
import org.apache.graphar.info.type.DataType;
import org.apache.graphar.info.yaml.EdgeYaml;
import org.apache.graphar.info.yaml.GraphYaml;
import org.apache.graphar.util.GeneralParams;
import org.yaml.snakeyaml.Yaml;

public class EdgeInfo {
    private final EdgeTriplet edgeTriplet;
    private final long chunkSize;
    private final long srcChunkSize;
    private final long dstChunkSize;
    private final boolean directed;
    private final URI baseUri;
    private final Map<AdjListType, AdjacentList> adjacentLists;
    private final PropertyGroups propertyGroups;
    private final VersionInfo version;

    public static EdgeInfoBuilder builder() {
        return new EdgeInfoBuilder();
    }

    public static final class EdgeInfoBuilder {
        private EdgeTriplet edgeTriplet;
        private long chunkSize;
        private long srcChunkSize;
        private long dstChunkSize;
        private boolean directed;
        private URI baseUri;
        private String prefix;
        private Map<AdjListType, AdjacentList> adjacentLists;
        private PropertyGroups propertyGroups;
        private VersionInfo version;

        private List<AdjacentList> adjacentListsAsListTemp;
        private List<PropertyGroup> propertyGroupsAsListTemp;

        private String srcType;
        private String edgeType;
        private String dstType;

        private EdgeInfoBuilder() {}

        public EdgeInfoBuilder srcType(String srcType) {
            this.srcType = srcType;
            return this;
        }

        public EdgeInfoBuilder edgeType(String edgeType) {
            this.edgeType = edgeType;
            return this;
        }

        public EdgeInfoBuilder dstType(String dstType) {
            this.dstType = dstType;
            return this;
        }

        public EdgeInfoBuilder edgeTriplet(String srcType, String edgeType, String dstType) {
            this.edgeTriplet = new EdgeTriplet(srcType, edgeType, dstType);
            return this;
        }

        public EdgeInfoBuilder edgeTriplet(EdgeTriplet edgeTriplet) {
            this.edgeTriplet = edgeTriplet;
            return this;
        }

        public EdgeInfoBuilder chunkSize(long chunkSize) {
            this.chunkSize = chunkSize;
            return this;
        }

        public EdgeInfoBuilder srcChunkSize(long srcChunkSize) {
            this.srcChunkSize = srcChunkSize;
            return this;
        }

        public EdgeInfoBuilder dstChunkSize(long dstChunkSize) {
            this.dstChunkSize = dstChunkSize;
            return this;
        }

        public EdgeInfoBuilder directed(boolean directed) {
            this.directed = directed;
            return this;
        }

        public EdgeInfoBuilder baseUri(URI baseUri) {
            this.baseUri = baseUri;
            return this;
        }

        public EdgeInfoBuilder prefix(String prefix) {
            this.prefix = prefix;
            return this;
        }

        public EdgeInfoBuilder addAdjacentList(AdjacentList adjacentList) {
            if (adjacentListsAsListTemp == null) {
                adjacentListsAsListTemp = new ArrayList<>();
            }
            adjacentListsAsListTemp.add(adjacentList);
            return this;
        }

        public EdgeInfoBuilder adjacentLists(List<AdjacentList> adjacentListsAsList) {
            if (adjacentListsAsListTemp == null) {
                adjacentListsAsListTemp = new ArrayList<>();
            }
            this.adjacentListsAsListTemp.addAll(adjacentListsAsList);
            return this;
        }

        public EdgeInfoBuilder adjacentLists(Map<AdjListType, AdjacentList> adjacentLists) {
            this.adjacentLists = adjacentLists;
            return this;
        }

        public EdgeInfoBuilder addPropertyGroup(PropertyGroup propertyGroup) {
            if (propertyGroupsAsListTemp == null) propertyGroupsAsListTemp = new ArrayList<>();
            propertyGroupsAsListTemp.add(propertyGroup);
            return this;
        }

        public EdgeInfoBuilder addPropertyGroups(List<PropertyGroup> propertyGroups) {
            if (propertyGroupsAsListTemp == null) propertyGroupsAsListTemp = new ArrayList<>();
            propertyGroupsAsListTemp.addAll(propertyGroups);
            return this;
        }

        public EdgeInfoBuilder propertyGroups(PropertyGroups propertyGroups) {
            this.propertyGroups = propertyGroups;
            return this;
        }

        public EdgeInfoBuilder version(String version) {
            this.version = VersionParser.getVersion(version);
            return this;
        }

        public EdgeInfoBuilder version(VersionInfo version) {
            this.version = version;
            return this;
        }

        public EdgeInfo build() {
            if (adjacentLists == null) {
                adjacentLists = new HashMap<>();
            }

            if (adjacentListsAsListTemp != null) {
                adjacentLists.putAll(
                        adjacentListsAsListTemp.stream()
                                .collect(
                                        Collectors.toUnmodifiableMap(
                                                AdjacentList::getType, Function.identity())));
            }

            if (propertyGroups == null && propertyGroupsAsListTemp != null) {
                propertyGroups = new PropertyGroups(propertyGroupsAsListTemp);
            } else if (propertyGroupsAsListTemp != null) {
                propertyGroups =
                        propertyGroupsAsListTemp.stream()
                                .map(propertyGroups::addPropertyGroupAsNew)
                                .filter(Optional::isPresent)
                                .map(Optional::get)
                                .reduce((first, second) -> second)
                                .orElse(new PropertyGroups(new ArrayList<>()));
            }

            if (edgeTriplet == null && srcType != null && edgeType != null && dstType != null) {
                edgeTriplet = new EdgeTriplet(srcType, edgeType, dstType);
            }

            if (edgeTriplet == null) {
                throw new IllegalArgumentException("Edge triplet is null");
            }

            if (propertyGroups == null) {
                throw new IllegalArgumentException("PropertyGroups is empty");
            }

            if (adjacentLists.isEmpty()) {
                throw new IllegalArgumentException("AdjacentLists is empty");
            }

            if (baseUri == null && prefix == null) {
                throw new IllegalArgumentException("baseUri and prefix cannot be both null");
            }

            if (baseUri != null && prefix != null && !URI.create(prefix).equals(baseUri)) {
                throw new IllegalArgumentException(
                        "baseUri and prefix conflict: baseUri="
                                + baseUri.toString()
                                + " prefix="
                                + prefix);
            }

            if (baseUri == null) {
                baseUri = URI.create(prefix);
            }

            return new EdgeInfo(this);
        }
    }

    private EdgeInfo(EdgeInfoBuilder builder) {
        this.edgeTriplet = builder.edgeTriplet;
        this.chunkSize = builder.chunkSize;
        this.srcChunkSize = builder.srcChunkSize;
        this.dstChunkSize = builder.dstChunkSize;
        this.directed = builder.directed;
        this.baseUri = builder.baseUri;
        this.adjacentLists = builder.adjacentLists;
        this.propertyGroups = builder.propertyGroups;
        this.version = builder.version;
    }

    public EdgeInfo(
            String srcType,
            String edgeType,
            String dstType,
            long chunkSize,
            long srcChunkSize,
            long dstChunkSize,
            boolean directed,
            URI baseUri,
            String version,
            List<AdjacentList> adjacentListsAsList,
            List<PropertyGroup> propertyGroupsAsList) {
        this(
                srcType,
                edgeType,
                dstType,
                chunkSize,
                srcChunkSize,
                dstChunkSize,
                directed,
                baseUri,
                VersionParser.getVersion(version),
                adjacentListsAsList,
                propertyGroupsAsList);
    }

    public EdgeInfo(
            String srcType,
            String edgeType,
            String dstType,
            long chunkSize,
            long srcChunkSize,
            long dstChunkSize,
            boolean directed,
            String prefix,
            String version,
            List<AdjacentList> adjacentListsAsList,
            List<PropertyGroup> propertyGroupsAsList) {
        this(
                srcType,
                edgeType,
                dstType,
                chunkSize,
                srcChunkSize,
                dstChunkSize,
                directed,
                URI.create(prefix),
                version,
                adjacentListsAsList,
                propertyGroupsAsList);
    }

    public EdgeInfo(
            String srcType,
            String edgeType,
            String dstType,
            long chunkSize,
            long srcChunkSize,
            long dstChunkSize,
            boolean directed,
            URI baseUri,
            VersionInfo version,
            List<AdjacentList> adjacentListsAsList,
            List<PropertyGroup> propertyGroupsAsList) {
        this.edgeTriplet = new EdgeTriplet(srcType, edgeType, dstType);
        this.chunkSize = chunkSize;
        this.srcChunkSize = srcChunkSize;
        this.dstChunkSize = dstChunkSize;
        this.directed = directed;
        this.baseUri = baseUri;
        this.adjacentLists =
                adjacentListsAsList.stream()
                        .collect(
                                Collectors.toUnmodifiableMap(
                                        AdjacentList::getType, Function.identity()));
        this.propertyGroups = new PropertyGroups(propertyGroupsAsList);
        this.version = version;
    }

    private EdgeInfo(
            EdgeTriplet edgeTriplet,
            long chunkSize,
            long srcChunkSize,
            long dstChunkSize,
            boolean directed,
            URI baseUri,
            String version,
            Map<AdjListType, AdjacentList> adjacentLists,
            PropertyGroups propertyGroups) {
        this(
                edgeTriplet,
                chunkSize,
                srcChunkSize,
                dstChunkSize,
                directed,
                baseUri,
                VersionParser.getVersion(version),
                adjacentLists,
                propertyGroups);
    }

    private EdgeInfo(
            EdgeTriplet edgeTriplet,
            long chunkSize,
            long srcChunkSize,
            long dstChunkSize,
            boolean directed,
            URI baseUri,
            VersionInfo version,
            Map<AdjListType, AdjacentList> adjacentLists,
            PropertyGroups propertyGroups) {
        this.edgeTriplet = edgeTriplet;
        this.chunkSize = chunkSize;
        this.srcChunkSize = srcChunkSize;
        this.dstChunkSize = dstChunkSize;
        this.directed = directed;
        this.baseUri = baseUri;
        this.adjacentLists = adjacentLists;
        this.propertyGroups = propertyGroups;
        this.version = version;
    }

    public static String concat(String srcLabel, String edgeLabel, String dstLabel) {
        return srcLabel
                + GeneralParams.regularSeparator
                + edgeLabel
                + GeneralParams.regularSeparator
                + dstLabel;
    }

    public Optional<EdgeInfo> addAdjacentListAsNew(AdjacentList adjacentList) {
        if (adjacentList == null || adjacentLists.containsKey(adjacentList.getType())) {
            return Optional.empty();
        }
        Map<AdjListType, AdjacentList> newAdjacentLists =
                Stream.concat(
                                adjacentLists.entrySet().stream(),
                                Map.of(adjacentList.getType(), adjacentList).entrySet().stream())
                        .collect(
                                Collectors.toUnmodifiableMap(
                                        Map.Entry::getKey, Map.Entry::getValue));
        return Optional.of(
                new EdgeInfo(
                        edgeTriplet,
                        chunkSize,
                        srcChunkSize,
                        dstChunkSize,
                        directed,
                        baseUri,
                        version,
                        newAdjacentLists,
                        propertyGroups));
    }

    public Optional<EdgeInfo> addPropertyGroupAsNew(PropertyGroup propertyGroup) {
        return propertyGroups
                .addPropertyGroupAsNew(propertyGroup)
                .map(
                        newPropertyGroups ->
                                new EdgeInfo(
                                        edgeTriplet,
                                        chunkSize,
                                        srcChunkSize,
                                        dstChunkSize,
                                        directed,
                                        baseUri,
                                        version,
                                        adjacentLists,
                                        newPropertyGroups));
    }

    public boolean hasAdjListType(AdjListType adjListType) {
        return adjacentLists.containsKey(adjListType);
    }

    public boolean hasProperty(String propertyName) {
        return propertyGroups.hasProperty(propertyName);
    }

    public boolean hasPropertyGroup(PropertyGroup propertyGroup) {
        return propertyGroups.hasPropertyGroup(propertyGroup);
    }

    public AdjacentList getAdjacentList(AdjListType adjListType) {
        // AdjListType will be checked in this method,
        // other methods which get adjacent list in this class should call this method first,
        // so we don't check AdjListType in other methods.
        checkAdjListTypeExist(adjListType);
        return adjacentLists.get(adjListType);
    }

    public int getPropertyGroupNum() {
        return propertyGroups.getPropertyGroupNum();
    }

    public PropertyGroup getPropertyGroup(String property) {
        return propertyGroups.getPropertyGroup(property);
    }

    public URI getPropertyGroupUri(PropertyGroup propertyGroup) {
        checkPropertyGroupExist(propertyGroup);
        return getBaseUri().resolve(propertyGroup.getBaseUri());
    }

    public URI getPropertyGroupChunkUri(PropertyGroup propertyGroup, long chunkIndex) {
        // PropertyGroup will be checked in getPropertyGroupPrefix
        return getPropertyGroupUri(propertyGroup).resolve("chunk" + chunkIndex);
    }

    public URI getAdjacentListUri(AdjListType adjListType) {
        return getBaseUri().resolve(getAdjacentList(adjListType).getBaseUri()).resolve("adj_list/");
    }

    public URI getAdjacentListChunkUri(AdjListType adjListType, long vertexChunkIndex) {
        return getAdjacentListUri(adjListType).resolve("chunk" + vertexChunkIndex);
    }

    public URI getOffsetUri(AdjListType adjListType) {
        return getAdjacentListUri(adjListType).resolve("offset/");
    }

    public URI getOffsetChunkUri(AdjListType adjListType, long vertexChunkIndex) {
        return getOffsetUri(adjListType).resolve("chunk" + vertexChunkIndex);
    }

    public URI getVerticesNumFileUri(AdjListType adjListType) {
        return getAdjacentListUri(adjListType).resolve("vertex_count");
    }

    public URI getEdgesNumFileUri(AdjListType adjListType, long vertexChunkIndex) {
        return getAdjacentListUri(adjListType).resolve("edge_count" + vertexChunkIndex);
    }

    public DataType getPropertyType(String propertyName) {
        return propertyGroups.getPropertyType(propertyName);
    }

    public boolean isPrimaryKey(String propertyName) {
        return propertyGroups.isPrimaryKey(propertyName);
    }

    public boolean isNullableKey(String propertyName) {
        return propertyGroups.isNullableKey(propertyName);
    }

    public String dump() {
        Yaml yaml = new Yaml(GraphYaml.getRepresenter(), GraphYaml.getDumperOptions());
        EdgeYaml edgeYaml = new EdgeYaml(this);
        return yaml.dump(edgeYaml);
    }

    public String getConcat() {
        return edgeTriplet.getConcat();
    }

    public String getSrcType() {
        return edgeTriplet.getSrcType();
    }

    public String getEdgeType() {
        return edgeTriplet.getEdgeType();
    }

    public String getDstType() {
        return edgeTriplet.getDstType();
    }

    public long getChunkSize() {
        return chunkSize;
    }

    public long getSrcChunkSize() {
        return srcChunkSize;
    }

    public long getDstChunkSize() {
        return dstChunkSize;
    }

    public boolean isDirected() {
        return directed;
    }

    public String getPrefix() {
        return baseUri.toString();
    }

    public URI getBaseUri() {
        return baseUri;
    }

    public VersionInfo getVersion() {
        return version;
    }

    public Map<AdjListType, AdjacentList> getAdjacentLists() {
        return adjacentLists;
    }

    public List<PropertyGroup> getPropertyGroups() {
        return propertyGroups.getPropertyGroupList();
    }

    private void checkAdjListTypeExist(AdjListType adjListType) {
        if (adjListType == null) {
            throw new IllegalArgumentException("The adjacency list type is null");
        }
        if (!adjacentLists.containsKey(adjListType)) {
            throw new IllegalArgumentException(
                    "The adjacency list type "
                            + adjListType
                            + " does not exist in the edge info "
                            + this.edgeTriplet.getConcat());
        }
    }

    private void checkPropertyGroupExist(PropertyGroup propertyGroup) {
        if (propertyGroup == null) {
            throw new IllegalArgumentException("Property group is null");
        }
        if (!hasPropertyGroup(propertyGroup)) {
            throw new IllegalArgumentException(
                    "Property group "
                            + propertyGroup
                            + " does not exist in the edge "
                            + getConcat());
        }
    }

    public boolean isValidated() {
        // Check if source type, edge type, or destination type is empty
        if (getSrcType() == null
                || getSrcType().isEmpty()
                || getEdgeType() == null
                || getEdgeType().isEmpty()
                || getDstType() == null
                || getDstType().isEmpty()) {
            return false;
        }

        // Check if chunk sizes are positive
        if (chunkSize <= 0 || srcChunkSize <= 0 || dstChunkSize <= 0) {
            return false;
        }

        // Check if prefix is null or empty
        if (baseUri == null || baseUri.toString().isEmpty()) {
            return false;
        }

        // Check if adjacent lists are empty
        if (adjacentLists.isEmpty()) {
            return false;
        }

        // Check if all adjacent lists are valid
        for (AdjacentList adjacentList : adjacentLists.values()) {
            if (adjacentList == null || !adjacentList.isValidated()) {
                return false;
            }
        }

        // Check if all property groups are valid and property names are unique
        Set<String> propertyNameSet = new HashSet<>();
        for (PropertyGroup pg : propertyGroups.getPropertyGroupList()) {
            if (pg == null || !pg.isValidated()) {
                return false;
            }

            for (Property p : pg.getPropertyList()) {
                if (propertyNameSet.contains(p.getName())) {
                    return false;
                }
                propertyNameSet.add(p.getName());
            }
        }
        return true;
    }

    private static class EdgeTriplet {
        private final String srcType;
        private final String edgeType;
        private final String dstType;

        public EdgeTriplet(String srcType, String edgeType, String dstType) {
            this.srcType = srcType;
            this.edgeType = edgeType;
            this.dstType = dstType;
        }

        public String getConcat() {
            return EdgeInfo.concat(srcType, edgeType, dstType);
        }

        @Override
        public String toString() {
            return getConcat();
        }

        public String getSrcType() {
            return srcType;
        }

        public String getEdgeType() {
            return edgeType;
        }

        public String getDstType() {
            return dstType;
        }
    }
}
